README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->

A Dart package that given an array of sized objects can return a card layout between three variations `horizontal, vertical and fan (rotated horizontal)`.

Example includes solution to be used via Flutter and via CustomPainter. The package is only responsible for delivering a List of coordinates and angle for each card, the presentation is up to the developer using the package.

![ezgif com-gif-maker](https://user-images.githubusercontent.com/864248/215085645-efacbb87-124b-4b20-bf36-55c722b5dbf7.gif)

---

## Getting started

The result of the layouts can be used with either Flutter widgets or via customPainter. The `/example` includes both ways.

## Usage

The package can layout cards in a `fan`, `horizontal` and `vertical` way.


```dart
List<Map<String, dynamic>> _cards = fanCards(flowCards, {
    "flow": "horizontal", // The layout direction (horizontal or vertical)
    "fanDirection": "N",
    "imagesUrl": "cards/",
    "spacing": 0.6,
    "radius": 200.0,
    "width": 80.0,
});

List<Map<String, dynamic>> _handCards = handCards(blockCards, {
    "flow": "horizontal",
    "spacing": -0.2,
    "width": 80.0,
});

List<Map<String, dynamic>> _columnCards = handCards(columnCards, {
    "flow": "vertical",
    "spacing": -0.2,
    "width": 80.0,
    "height": 120.0,
});
```

## Additional information

This package is a direct port of https://github.com/richardschneider/cardsJS 


CHANGELOG.md
## 1.0.1

* fixed formatting issues
## 1.0.0

* Initial release


test/playing_cards_layouts_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:playing_cards_layouts/playing_cards_layouts.dart';

void main() {
  test('adds one to input values', () {
    final cards = fanCards([], {});
    expect(cards.length, 0);
  });
}


lib/utils/math.dart
import 'dart:math';

double degreesToRadians(double degrees) {
  return degrees * (pi / 180);
}

double radiansToDegrees(double radians) {
  return radians * (180 / pi);
}

List<double> getRotatedDimensions(
  double angleInDegrees,
  double width,
  double height,
) {
  double angle = angleInDegrees * pi / 180;
  double sinValue = sin(angle);
  double cosValue = cos(angle);
  double x1 = cosValue * width;
  double y1 = sinValue * width;
  double x2 = -sinValue * height;
  double y2 = cosValue * height;
  double x3 = cosValue * width - sinValue * height;
  double y3 = sinValue * width + cosValue * height;
  List<double> sortX = <double>[0.0, x1, x2, x3];
  sortX.sort();
  List<double> sortY = <double>[0.0, y1, y2, y3];
  sortY.sort();
  double minX = sortX.first;
  double maxX = sortX.last;
  double minY = sortY.first;
  double maxY = sortY.last;

  return [((maxX - minX)).floorToDouble(), ((maxY - minY)).floorToDouble()];
}

List<double> rotatePointInBox(
  double x,
  double y,
  double angle,
  double width,
  double height,
) {
  double angleValue = degreesToRadians(angle);
  double centerX = width / 2.0;
  double centerY = height / 2.0;
  double dx = x - centerX;
  double dy = y - centerY;
  double dist = sqrt(dx * dx + dy * dy);
  double a = atan2(dy, dx) + angleValue;
  double dx2 = cos(a) * dist;
  double dy2 = sin(a) * dist;

  return [dx2 + centerX, dy2 + centerY];
}

List<Map<String, dynamic>> calculateCoords(
    int numCards,
    double arcRadius,
    double cardWidth,
    double cardHeight,
    String direction,
    double cardSpacing,
    Map<String, dynamic> box) {
  // The separation between the cards, in terms of rotation around the circle's origin
  var anglePerCard =
      radiansToDegrees(atan((cardWidth * cardSpacing) / arcRadius));
  String directionOption = direction.toUpperCase();
  int angleOffset = {"N": 270, "S": 90, "E": 0, "W": 180}[directionOption]!;

  var startAngle = angleOffset - 0.5 * anglePerCard * (numCards - 1);

  List<Map<String, dynamic>> coords = [];
  int i = 0;
  double minX = 99999.0;
  double minY = 99999.0;
  double maxX = -minX;
  double maxY = -minY;
  for (i = 0; i < numCards; i++) {
    double degrees = startAngle + anglePerCard * i;

    double radians = degreesToRadians(degrees);
    var x = cardWidth / 2 + cos(radians) * arcRadius;
    var y = cardHeight / 2 + sin(radians) * arcRadius;

    minX = min(minX, x);
    minY = min(minY, y);
    maxX = max(maxX, x);
    maxY = max(maxY, y);

    coords.add({"x": x, "y": y, "angle": degrees});
  }

  List<double> rotatedDimensions =
      getRotatedDimensions(coords[0]["angle"], cardWidth, cardHeight);

  double offsetX = 0.0;
  double offsetY = 0.0;

  if (directionOption == "N") {
    offsetX = minX * -1;
    offsetX += (rotatedDimensions[0] - cardWidth) / 2;

    offsetY = minY * -1;
  } else if (directionOption == "S") {
    offsetX = minX * -1;
    offsetX += (rotatedDimensions[0] - cardWidth) / 2;

    offsetY = (minY + (maxY - minY)) * -1;
  } else if (directionOption == "W") {
    offsetY = minY * -1;
    offsetY += (rotatedDimensions[1] - cardHeight) / 2;

    offsetX = minX * -1;
    offsetX +=
        cardHeight - rotatePointInBox(0, 0, 270, cardWidth, cardHeight)[1];
  } else if (directionOption == "E") {
    offsetY = minY * -1;
    offsetY += (rotatedDimensions[1] - cardHeight) / 2;

    offsetX = arcRadius * -1;
    offsetX -=
        cardHeight - rotatePointInBox(0, 0, 270, cardWidth, cardHeight)[1];
    //offsetX -= ?????;    // HELP! Needs to line up with yellow line!
  }

  for (int i = 0; i < coords.length; i++) {
    coords[i]["x"] += offsetX;
    coords[i]["x"] = (coords[i]["x"]);

    coords[i]["y"] += offsetY;
    coords[i]["y"] = (coords[i]["y"]);

    coords[i]["angle"] = coords[i]["angle"];
  }

  box["width"] = coords[numCards - 1]["x"] + cardWidth;
  box["height"] = coords[numCards - 1]["y"] + cardHeight;

  return coords;
}


example/lib/card_builder.dart
import 'dart:core';
import 'dart:math';
import 'package:flutter/material.dart';
import 'dart:ui' as ui;
import 'package:vector_math/vector_math.dart' as vectorMath;

class CardBuilder extends CustomPainter {
  List<Map<String, dynamic>> points = [];
  var index = 0;
  AnimationController? controller;
  Canvas? canvas;
  int currentTime = 0;
  int fps = 24;
  int timeDecay = 0;
  double width = 100;
  double height = 100;
  double _radius = 10;
  Color shadowColor = Color.fromRGBO(0, 0, 0, 0.3);
  Function? update;
  Paint _paint = Paint();
  ui.BlendMode? blendMode = ui.BlendMode.src;
  List<Map<String, dynamic>> cardCoords = [];
  List cards = [];
  BlendMode _blendMode = BlendMode.dstATop;
  //

  /// Constructor
  CardBuilder({
    /// <-- The animation controller
    required this.controller,

    /// <-- The delay until the animation starts
    required this.cards,
    radius,
    shadowColor,

    /// <-- The particles blend mode (default: BlendMode.src)
    blendMode,
  }) : super(repaint: controller) {
    /// the delay in ms based on desired fps
    timeDecay = (1 / fps * 1000).round();
    _paint = Paint()
      ..strokeCap = StrokeCap.round
      ..isAntiAlias = true
      ..color = Colors.yellow
      ..strokeWidth = 10.0
      ..strokeJoin = StrokeJoin.round
      ..style = PaintingStyle.fill;

    _radius = radius ?? 10.0;
    _blendMode = blendMode ?? BlendMode.dstATop;
    this.shadowColor = shadowColor ?? Color.fromRGBO(0, 0, 0, 0.3);
  }

  @override
  void paint(Canvas canvas, Size size) {
    this.canvas = canvas;

    paintImage(canvas, size);
  }

  void paintImage(Canvas canvas, Size size) async {
    draw(canvas, size);
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return false;
  }

  void draw(Canvas canvas, Size size) {
    ///
    ////// draw cards
    drawCards();
  }

  void drawCards() {
    int i = 0;
    for (var card in cards) {
      drawRRect(card, i);
      i += 1;
    }
  }

  void drawRRect(Map<String, dynamic> card, int index) {
    if (canvas != null) {
      //print(card);
      if (card["coords"] != null) {
        var path = Path();

        _paint.color = card["card"]["color"];
        Paint shadowPaint = Paint()
          ..color = shadowColor
          ..maskFilter = MaskFilter.blur(BlurStyle.normal, 20);
        /////////////////////////////////
        double angle = 0;
        double translateX = 0.0;
        double translateY = 0.0;
        double cardWidth = card["card"]["width"];
        double cardHeight = card["card"]["height"];
        if (card["coords"]["angle"] != null) {
          angle = card["coords"]["angle"].toDouble();
          final double r =
              sqrt(cardWidth * cardWidth + cardHeight * cardHeight) / 2;
          final alpha = atan(cardHeight / cardWidth);
          final beta = alpha + angle * (pi / 180);
          final shiftY = r * sin(beta);
          final shiftX = r * cos(beta);
          translateX = cardWidth / 2 - shiftX;
          translateY = cardHeight / 2 - shiftY;
        }
        //
        updateCanvas(canvas!, card["coords"]["x"] + translateX,
            card["coords"]["y"] + translateY, angle * (pi / 180), () {
          //path = Path();

          // add shadow
          Rect rectS = Rect.fromLTWH(0, 6, card["card"]["width"].toDouble(),
              card["card"]["height"].toDouble());
          canvas!.drawRRect(
              RRect.fromRectAndRadius(rectS, Radius.circular(_radius)),
              shadowPaint);

          Rect rect = Rect.fromLTWH(0, 0, card["card"]["width"].toDouble(),
              card["card"]["height"].toDouble());
          path.addRRect(
              RRect.fromRectAndRadius(rect, Radius.circular(_radius)));
          canvas!.drawRRect(
              RRect.fromRectAndRadius(rect, Radius.circular(_radius)), _paint);
          //
        }, translate: true);
      }
    }
  }

  void updateCanvas(Canvas canvas, double? x, double? y, double? rotate,
      VoidCallback callback,
      {bool translate = false}) {
    double _x = x ?? 0;
    double _y = y ?? 0;

    canvas.save();

    if (translate) {
      //canvas.translate(_x, _y);
    }

    if (rotate != null) {
      canvas.translate(_x, _y);
      canvas.rotate(rotate);
      //canvas.translate(-_x, -_y);
    }
    callback();
    canvas.restore();
  }
}


lib/playing_cards_layouts.dart
import 'package:playing_cards_layouts/utils/math.dart';

/// fan cards layout
List<Map<String, dynamic>> fanCards(
    List<Map<String, dynamic>> cards, Map<String, dynamic> options) {
  var n = cards.length;
  if (n == 0) {
    return [];
  }

  double width = options["width"] ?? 90.0; // hack: for a hidden hand
  double height =
      cards[0]["height"].toDouble() ?? (width * 1.4); // hack: for a hidden hand

  Map<String, dynamic> box = {};
  List<Map<String, dynamic>> finalCards = [];
  var coords = calculateCoords(n, options["radius"], width, height,
      options["fanDirection"], options["spacing"], box);

  var i = 0;
  for (var coord in coords) {
    var card = cards[i];
    finalCards.add({"card": card, "coords": coord});

    i += 1;
  }

  return finalCards;
}

/// hand cards layout (horizontal, vertical)
List<Map<String, dynamic>> handCards(
    List<Map<String, dynamic>> cards, Map<String, dynamic> options) {
  var n = cards.length;
  if (n == 0) {
    return [];
  }

  List<Map<String, dynamic>> finalCards = [];

  var width =
      options["width"] ?? cards[0]["width"] ?? 70; // hack: for a hidden hand
  var height =
      cards[0]["height"] ?? (width * 1.4).floor(); // hack: for a hidden hand
  int counter = 0;
  for (var card in cards) {
    if (options["flow"] == 'vertical') {
      finalCards.add({
        "card": card,
        "coords": {
          "x": 0.0,
          "y": ((height * counter) + ((options["spacing"] * height) * counter)),
        },
      });
    } else if (options["flow"] == 'horizontal') {
      finalCards.add({
        "card": card,
        "coords": {
          "x": ((width * counter) + ((options["spacing"] * width) * counter)),
          "y": 0.0,
        },
      });
    }

    counter += 1;
  }
  return finalCards;
}


example/lib/main.dart
import 'dart:async';
import 'dart:math';

import 'package:example/card_builder.dart';
import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';
import 'dart:ui' as ui;

import 'package:playing_cards_layouts/playing_cards_layouts.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> with TickerProviderStateMixin {
  late AnimationController _controller;
  BoxConstraints? viewportConstraints;
  List<Map<String, dynamic>> flowCards = [];
  List<Map<String, dynamic>> blockCards = [];
  List<Map<String, dynamic>> columnCards = [];
  double visible = 0.0;
  @override
  void initState() {
    super.initState();

    _controller =
        AnimationController(vsync: this, duration: const Duration(seconds: 1));

    SchedulerBinding.instance.addPostFrameCallback((_) {
      //_controller.repeat();
      generateCards();
    });
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  /// Generate the maze
  void generateCards() {
    for (int i = 0; i < 5; i++) {
      flowCards.add({
        "color": Color.fromARGB(
            255, Colors.purple.red, Colors.purple.green, Colors.purple.blue),
        "id": UniqueKey().toString(),
        "width": 80.0,
        "height": 120.0,
      });

      blockCards.add({
        "color": Color.fromARGB(
            255, Colors.blue.red, Colors.blue.green, Colors.blue.blue),
        "id": UniqueKey().toString(),
        "width": 80.0,
        "height": 120.0,
      });

      columnCards.add({
        "color": Color.fromARGB(
            255, Colors.green.red, Colors.green.green, Colors.green.blue),
        "id": UniqueKey().toString(),
        "width": 80.0,
        "height": 120.0,
      });
    }

    List<Map<String, dynamic>> _cards = fanCards(flowCards, {
      "flow": "horizontal", // The layout direction (horizontal or vertical)
      "fanDirection": "n",
      "imagesUrl": "cards/",
      "spacing": 0.6,
      "radius": 200.0,
      "width": 80.0,
    });

    List<Map<String, dynamic>> _handCards = handCards(blockCards, {
      "flow": "horizontal",
      "spacing": -0.2,
      "width": 80.0,
    });

    List<Map<String, dynamic>> _columnCards = handCards(columnCards, {
      "flow": "vertical",
      "spacing": -0.2,
      "width": 80.0,
      "height": 120.0,
    });

    flowCards.clear();
    blockCards.clear();
    columnCards.clear();
    setState(() {
      flowCards = _cards;
      blockCards = _handCards;
      columnCards = _columnCards;
    });
  }

  List<Widget> getCards(List<Map<String, dynamic>> sourceCards, String mode) {
    List<Widget> cards = [];
    cards.add(Container());
    int counter = 0;
    sourceCards.asMap().forEach((key, value) {
      double? angle = value["coords"]["angle"];
      if (angle != null) {
        angle = (angle - 90) * (pi / 180);
      }

      cards.add(
        Positioned(
          left: value["coords"]["x"].toDouble(),
          top: value["coords"]["y"].toDouble(),
          child: GestureDetector(
            behavior: HitTestBehavior.translucent,
            onTapDown: (details) => onTap(context, details, mode, key),
            child: AbsorbPointer(
              child: Transform.translate(
                offset: const Offset(0, 0),
                child: Transform.rotate(
                  angle: angle ?? 0.0,
                  child: FadeTransition(
                    opacity: animateOpacity(counter),
                    child: Container(
                      height: value["card"]["height"],
                      width: value["card"]["width"],
                      color: Colors.transparent,
                      child: Container(
                        decoration: BoxDecoration(
                          color: value["card"]["color"],
                          borderRadius: BorderRadius.all(
                            Radius.circular(10.0),
                          ),
                          boxShadow: [
                            BoxShadow(
                              color: Colors.black.withOpacity(0.19),
                              spreadRadius: 4,
                              blurRadius: 6,
                              offset:
                                  Offset(0, 3), // changes position of shadow
                            ),
                          ],
                        ),
                      ),
                    ),
                  ),
                ),
              ),
            ),
          ),
        ),
      );

      counter += 1;
    });

    return cards;
  }

  void randomize() {
    generateCards();
  }

  void onTap(
      BuildContext context, TapDownDetails details, String mode, int index) {
    // find the card
    var card = mode == "horizontal" ? blockCards : columnCards;
    double dx = details.globalPosition.dx - 100;
    double dy = details.globalPosition.dy;

    if (mode == "vertical") {
      dx = details.globalPosition.dx - 100;
      dy = details.globalPosition.dy - 600;
    }

    var cardWidth = card[0]["card"]["width"];
    var cardHeight = card[0]["card"]["height"];
    var offsetX = cardWidth * 0.2;
    var offsetY = cardHeight * 0.2;
    var indexX = (dx / (cardWidth - offsetX)).floor();
    var indexY = (dy / (cardHeight - offsetY)).floor();

    // small quality of life enhancement
    if (mode == "horizontal_rotated") {
      indexX = index;
    }

    if (indexX >= card.length) {
      indexX = card.length - 1;
    }
    if (indexX < 0) {
      indexX = 0;
    }

    if (indexY >= card.length) {
      indexY = card.length - 1;
    }
    if (indexY < 0) {
      indexY = 0;
    }

    if (mode == "horizontal" || mode == "horizontal_rotated") {
      animateCard(indexX, () => {setState(() => {})}, mode);
    } else {
      animateCard(indexY, () => {setState(() => {})}, mode);
    }
  }

  Animation<double> animateOpacity(int index) {
    AnimationController controller = AnimationController(
        duration: const Duration(milliseconds: 1000), vsync: this);
    Animation<double> tween =
        Tween<double>(begin: 0.0, end: 1.0).animate(CurvedAnimation(
      parent: _controller,
      curve: Interval(
        0.1 * index,
        1.0,
        curve: Curves.ease,
      ),
    ));

    tween.addListener(() => {
          //setState(() => {}),
        });

    Future.delayed(const Duration(milliseconds: 1000), () {
      _controller.forward();
    });

    return tween;
  }

  // animate the card
  void animateCard(int cardIndex, Function updateFn, String mode) {
    var card = mode == "horizontal"
        ? blockCards[cardIndex]
        : mode == "horizontal_rotated"
            ? flowCards[cardIndex]
            : columnCards[cardIndex];
    if (mode == "horizontal_rotated") {
      card = flowCards[cardIndex];
    }
    var currentY = card["coords"]["y"];
    var currentX = card["coords"]["x"];
    double from = 0;
    double to = -10;

    if (mode == "horizontal") {
      from = 0;
      to = -10;
    }

    if (mode == "horizontal_rotated") {
      if (card["coords"]["originalY"] == null) {
        card["coords"]["originalY"] = card["coords"]["y"];
      }
      from = card["coords"]["y"];
      to = (card["coords"]["y"] - 10);
    }

    if (currentY < 0) {
      from = currentY;
      to = 0;
    }

    if (mode == "horizontal_rotated") {
      if (currentY != card["coords"]["originalY"]) {
        to = card["coords"]["originalY"];
      }
    }

    if (mode == "vertical") {
      from = 0;
      to = 10;

      if (currentX > 0) {
        from = currentX;
        to = 0;
      }
    }

    AnimationController controller = AnimationController(
        duration: const Duration(milliseconds: 100), vsync: this);
    final Animation<double> curve =
        CurvedAnimation(parent: controller, curve: Curves.easeOut);
    Animation<double> tween = Tween(begin: from, end: to).animate(curve);

    tween.addListener(() => {
          if (mode == "horizontal" || mode == "horizontal_rotated")
            {
              card["coords"]["y"] = tween.value,
            }
          else
            {
              card["coords"]["x"] = tween.value,
            },
          updateFn(),
        });
    controller.forward();
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
        backgroundColor: Colors.white,
        bottomNavigationBar: Padding(
          padding: const EdgeInsets.all(8.0),
          child: Container(
              padding: const EdgeInsets.only(left: 8.0, right: 8.0),
              decoration: BoxDecoration(
                  borderRadius: BorderRadius.circular(50.0),
                  color: const ui.Color.fromARGB(255, 0, 0, 0)),
              child: Padding(
                padding: const EdgeInsets.all(8.0),
                child: Column(
                  mainAxisSize: MainAxisSize.min,
                  children: <Widget>[
                    Row(
                      mainAxisAlignment: MainAxisAlignment.spaceBetween,
                      children: const <Widget>[],
                    ),
                  ],
                ),
              )),
        ),
        body: LayoutBuilder(builder:
            (BuildContext context, BoxConstraints viewportConstraints) {
          this.viewportConstraints = viewportConstraints;
          return Stack(children: [
            Positioned(
              top: 100,
              left: 100,
              child: SizedBox(
                width: flowCards.length * 80,
                height: 300,
                child: Stack(
                  fit: StackFit.passthrough,
                  clipBehavior: Clip.none,
                  children: getCards(flowCards, "horizontal_rotated"),
                ),
              ),
              // Transform.translate(
              //     offset: const Offset(0, 0),
              //     child: CustomPaint(
              //       size: const ui.Size(200, 400),
              //       key: UniqueKey(),
              //       isComplex: true,
              //       painter: CardBuilder(
              //         controller: _controller,
              //         cards: flowCards,
              //         //solution: this.mazeSolution,
              //         // width: viewportConstraints.maxWidth,
              //         // height: viewportConstraints.maxHeight,
              //       ),
              //       child: Container(constraints: BoxConstraints(maxWidth: viewportConstraints.maxWidth, maxHeight: viewportConstraints.maxHeight)),
              //     )),
            ),
            Positioned(
              top: 400,
              left: 100,
              child: SizedBox(
                width: blockCards.length * 80,
                height: 120,
                child: Stack(
                  fit: StackFit.passthrough,
                  clipBehavior: Clip.none,
                  children: getCards(blockCards, "horizontal"),
                ),
              ),
            ),
            Positioned(
              top: 600,
              left: 100,
              child: SizedBox(
                width: 80,
                height: columnCards.length * 120,
                child: Stack(
                  fit: StackFit.passthrough,
                  clipBehavior: Clip.none,
                  children: getCards(columnCards, "vertical"),
                ),
              ),
            ),
            Positioned(
              top: viewportConstraints.maxHeight - 100,
              left: viewportConstraints.maxWidth - 100,
              child: FloatingActionButton(
                onPressed: randomize,
                backgroundColor: Colors.green,
                child: const Icon(Icons.refresh),
              ),
            ),
          ]);

          //);
        }));
  }
}
